### ISSUE TITLE
autoregressive wrapper generate should accommodate variable lengthed prefixes

### ISSUE THREAD

**Issue by LouChao98** [lucidrains/x-transformers#185]
https://github.com/lucidrains/x-transformers/blob/a3373cb9f7ead1db1905eab420dd2c80e263e054/x_transformers/x_transformers.py#L1195

Here we need to slice the mask as well.

**Comment by lucidrains** [comment]
@LouChao98 oh, when would you need a mask for decoder training?

**Comment by lucidrains** [comment]
@LouChao98 i'm just going to disallow it until a legitimate use case comes up https://github.com/lucidrains/x-transformers/commit/a651c19ac108e45f5a8e8ab9a74d04a41988aaf2

**Comment by LouChao98** [comment]
To prevent access to left padding tokens, a mask is required for batch decoding with different prefixes.

**Comment by LouChao98** [comment]
Oh, my mistake.  The issue is here

https://github.com/lucidrains/x-transformers/blob/2d1aca2df5c282170b168cf5e5ab04cda18df077/x_transformers/x_transformers.py#L914-L916

This should be disabled if using cache. No need to slice mask.

**Comment by lucidrains** [comment]
@LouChao98 how are you doing your positional embedding in the presence of left padding tokens?

**Comment by lucidrains** [comment]
@LouChao98 actually, i can fix your issue in another way by tomorrow's end, without masking of left padding. i haven't considered variable lengthed prefixes in the current implementation

**Comment by lucidrains** [comment]
@LouChao98 i'm kind of in this mindspace anyways; currently stuck on how to approach batched speculative decoding (where i am actually considering left padding tokens and facing positional embedding issues)

**Comment by lucidrains** [comment]
the solution for you would be passing in the prefix lengths, and incrementing it and indexing out the correct logit on each iteration

**Comment by lucidrains** [comment]
actually maybe you are right with the left padding tokens, I can't think of a clean way to manage the kv cache otherwise

ok let me meditate on this tomorrow morning over a cup of coffee or two

**Comment by LouChao98** [comment]
If we can assume the mask is continuous (left padded and right padded), I think the only change to pos emb is about absolute pos emb. We can use the cumsum of the mask instead of `torch.arange` to get the correct pos emb.

If not use a mask or padding, we may use a block-wise attention. I know a paper implement a more complexicated attn (hash-sparse attn), achieving a near flash-attn speed: https://openreview.net/pdf?id=BQEaklwG9P

**Comment by lucidrains** [comment]
@LouChao98 yea no worries, i can fix the variable prefix length problem, and in turn also make a dent in the batched spec decoding problem

yea, that hash sparse attention paper isn't great, just use flash attention 2

**Comment by lucidrains** [comment]
@LouChao98 do you want to check out 1.22.8? you will simply have to pass in `prompt_lens` at `.generate` and it should automatically take care of all the masking, positional offsetting, etc (all while maintaining kv cache)

**Comment by lucidrains** [comment]
ok, this was actually very helpful, as it leads up to batched spec decoding, which is much more difficult (but made simpler after crafting the align function)

**Comment by LouChao98** [comment]
I doubt the need of realign:

https://github.com/lucidrains/x-transformers/blob/98d750f7c607563f3f00edc319a8c2d3a36e17ef/x_transformers/autoregressive_wrapper.py#L249-L252

Just `out = out[:, t:]` can extract the generate texts. 

Besides, because `prompt_lens` is not updated, realign + slice will always get zeros.

**Comment by LouChao98** [comment]
I also suggest to remove `seq_start_pos` in `RotaryEmbedding` because it do not affect the result of self attn but confuse readers.

**Comment by lucidrains** [comment]
> I doubt the need of realign:
> 
> https://github.com/lucidrains/x-transformers/blob/98d750f7c607563f3f00edc319a8c2d3a36e17ef/x_transformers/autoregressive_wrapper.py#L249-L252
> 
> Just `out = out[:, t:]` can extract the generate texts.
> 
> Besides, because `prompt_lens` is not updated, realign + slice will always get zeros.

yes you are right, made the change

and yes, you are also right about rotary, it should be shift invariant, i think? i'll need to revisit the math

**Comment by lucidrains** [comment]
@LouChao98 ok, i've reverted the rotary embedding offsetting

do you want to give it one more review?

**Comment by LouChao98** [comment]
I found a bug in `AbsolutePositionalEmbedding` and `ScaledSinusoidalEmbedding`. I am sorry for not checking the abs emb parts in previous.

https://github.com/lucidrains/x-transformers/blob/77c8464785a5e261c52c1ddd7a48cc53c30f5528/x_transformers/x_transformers.py#L218-L219

The offset should be applied as follows, because during generation, the inputs are left padded.
```
pos = (pos - offset[..., None]).relu()
```

Everything else looks great.

**Comment by lucidrains** [comment]
@LouChao98 🤦 ok, yes, [should be fixed](https://github.com/lucidrains/x-transformers/commit/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9)

**Comment by LouChao98** [comment]
Hah, great. Thanks for your effort. I think we can close this now.

**Comment by lucidrains** [comment]
@LouChao98 yea no problem, thanks for opening the issue. really led somewhere nice
